src/Couchbase.Lite.Shared/API/Sync/Replicator.cs

```diff
@@ -262,10 +262,40 @@ private static void StatusChangedCallback(C4Replicator* repl, C4ReplicatorStatus
             });
         }
 
+        [MonoPInvokeCallback(typeof(C4ReplicatorPushFilterFunction))]
+        private static bool PushFilterCallback(FLSlice docID, FLDict* dict, void* context)
+        { 
+            var replicator = GCHandle.FromIntPtr((IntPtr)context).Target as Replicator;
+            var docIDStr = docID.CreateString();
+            return replicator.PushFilterCallback(docIDStr, dict);
+        }
+
+        bool PushFilterCallback(string docID, FLDict* dict)
+        {
+            if (Config.PushFilter == null)
+                return true;
+            var doc = Config.Database.GetDocument(docID);
+            var f = Config.PushFilter;
+            var v = f(doc);
+            return v;
+        }
+
         [MonoPInvokeCallback(typeof(C4ReplicatorValidationFunction))]
-        private static bool ValidateCallback(string docID, IntPtr body, object context)
+        private static bool PullValidateCallback(FLSlice docID, FLDict* dict, void* context)
         {
-            return true;
+            var replicator = GCHandle.FromIntPtr((IntPtr)context).Target as Replicator;
+            var docIDStr = docID.CreateString();
+            return replicator.PullValidateCallback(docIDStr, dict);
+        }
+
+        bool PullValidateCallback(string docID, FLDict* dict)
+        {
+            if (Config.PullFilter == null)
+                return true;
+            var doc = Config.Database.GetDocument(docID);
```

Comment:
I'm not sure this is the way that was intended.  What is coming with "dict"?  If it is the document properties then those should be used instead.  Otherwise there will be a database call for every item in the pull replication which will slow things down a lot (ditto for the push callback).

Reply:
I actually don't know how to get info out FLDict.

Reply:
You could use FLValueConverter to convert it to a .NET dictionary and examine in the debugger.  Or ask Jens / Pasin what this dict is supposed to contain.

Reply:
remove Config.PullFilter == null because it will never get to here is Config.PullFilter == null